{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e8439cbd",
   "metadata": {
    "papermill": {
     "duration": 0.018345,
     "end_time": "2023-05-04T09:45:23.621293",
     "exception": false,
     "start_time": "2023-05-04T09:45:23.602948",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "GOOD = 'v1.0.4'  # where script exits 0\n",
    "BAD = 'v1.5.3'  # where script exists 1\n",
    "SCRIPT = \"\"\"\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "    n = 2 ** i\n",
    "\n",
    "    print(f\"Trying n = 2 ** {i} '{n}'...\")\n",
    "\n",
    "    df = pd.DataFrame([['A', 14], ['A', n]], columns=['gb', 'val'])\n",
    "\n",
    "    gb_sum = df.groupby('gb').sum().values[0][0]\n",
    "    df_sum = df.sum().values[1]\n",
    "    \n",
    "    assert gb_sum != df_sum\n",
    "\n",
    "    del df\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0577b863",
   "metadata": {
    "papermill": {
     "duration": 33.449567,
     "end_time": "2023-05-04T09:45:57.076581",
     "exception": false,
     "start_time": "2023-05-04T09:45:23.627014",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import shlex\n",
    "import shutil\n",
    "\n",
    "if os.path.exists('pandas-dev'):\n",
    "    shutil.rmtree('pandas-dev')\n",
    "_ = !git clone https://github.com/pandas-dev/pandas.git pandas-dev"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "453f012c",
   "metadata": {
    "papermill": {
     "duration": 0.016485,
     "end_time": "2023-05-04T09:45:57.097625",
     "exception": false,
     "start_time": "2023-05-04T09:45:57.081140",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "with open('pandas-dev/t.py', 'w') as fd:\n",
    "    fd.write(SCRIPT)\n",
    "    \n",
    "VERSIONEER = \"\"\"\n",
    "import os\n",
    "has_versioneer = False\n",
    "with open('requirements-dev.txt') as fd:\n",
    "    for line in fd:\n",
    "        if 'versioneer' in line:\n",
    "            has_versioneer = True\n",
    "            cython_rqt = line\n",
    "print('has versioneer: ', has_versioneer)\n",
    "os.system(f'pip install {line}')\n",
    "if has_versioneer:\n",
    "    os.system('pip install versioneer[toml]')\n",
    "else:\n",
    "    os.system('pip uninstall versioneer -y')\n",
    "\"\"\"\n",
    "    \n",
    "with open('pandas-dev/install_versioneer.py', 'w') as fd:\n",
    "    fd.write(VERSIONEER)\n",
    "\n",
    "BISECT = \"\"\"\n",
    "python install_versioneer.py\n",
    "python setup.py build_ext -j 1\n",
    "pip install -e . --no-build-isolation --no-use-pep517\n",
    "python t.py\n",
    "\"\"\"\n",
    "\n",
    "with open('pandas-dev/bisect.sh', 'w') as fd:\n",
    "    fd.write(BISECT)\n",
    "    \n",
    "SETUP_BISECT = f\"\"\"\n",
    "git bisect start\n",
    "git bisect good {GOOD}\n",
    "git bisect bad {BAD}\n",
    "\"\"\"\n",
    "\n",
    "with open('pandas-dev/setup_bisect.sh', 'w') as fd:\n",
    "    fd.write(SETUP_BISECT)\n",
    "    \n",
    "CHECKOUT_BAD = f\"\"\"\n",
    "git checkout {BAD}\n",
    "\"\"\"\n",
    "\n",
    "with open('pandas-dev/checkout_bad.sh', 'w') as fd:\n",
    "    fd.write(CHECKOUT_BAD)\n",
    "\n",
    "CHECKOUT_GOOD = f\"\"\"\n",
    "git checkout {GOOD}\n",
    "\"\"\"\n",
    "\n",
    "with open('pandas-dev/checkout_good.sh', 'w') as fd:\n",
    "    fd.write(CHECKOUT_GOOD)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9a0e9d99",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\ngit checkout v1.5.3\\n'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CHECKOUT_BAD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ae221ac2",
   "metadata": {
    "papermill": {
     "duration": 1.153772,
     "end_time": "2023-05-04T09:45:58.256233",
     "exception": false,
     "start_time": "2023-05-04T09:45:57.102461",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Note: switching to 'v1.5.3'.\r\n",
      "\r\n",
      "You are in 'detached HEAD' state. You can look around, make experimental\r\n",
      "changes and commit them, and you can discard any commits you make in this\r\n",
      "state without impacting any branches by switching back to a branch.\r\n",
      "\r\n",
      "If you want to create a new branch to retain commits you create, you may\r\n",
      "do so (now or later) by using -c with the switch command. Example:\r\n",
      "\r\n",
      "  git switch -c <new-branch-name>\r\n",
      "\r\n",
      "Or undo this operation with:\r\n",
      "\r\n",
      "  git switch -\r\n",
      "\r\n",
      "Turn off this advice by setting config variable advice.detachedHead to false\r\n",
      "\r\n",
      "HEAD is now at 2e218d1098 RLS: 1.5.3\r\n"
     ]
    }
   ],
   "source": [
    "!cd pandas-dev && bash checkout_bad.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ec36b9b1",
   "metadata": {
    "papermill": {
     "duration": 1.106048,
     "end_time": "2023-05-04T09:45:59.367091",
     "exception": false,
     "start_time": "2023-05-04T09:45:58.261043",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[33mcommit 2e218d10984e9919f0296931d92ea851c6a6faf5\u001b[m\u001b[33m (\u001b[m\u001b[1;36mHEAD\u001b[m\u001b[33m, \u001b[m\u001b[1;33mtag: v1.5.3\u001b[m\u001b[33m)\u001b[m\r\n",
      "Author: Pandas Development Team <pandas-dev@python.org>\r\n",
      "Date:   Thu Jan 19 10:26:43 2023 +0700\r\n",
      "\r\n",
      "    RLS: 1.5.3\r\n"
     ]
    }
   ],
   "source": [
    "!cd pandas-dev && git log -n 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "eba1fd03",
   "metadata": {
    "papermill": {
     "duration": 12.78244,
     "end_time": "2023-05-04T09:46:12.154321",
     "exception": false,
     "start_time": "2023-05-04T09:45:59.371881",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting cython\n",
      "  Using cached Cython-3.0.0-cp37-cp37m-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.5 MB)\n",
      "Collecting numpy\n",
      "  Using cached numpy-1.21.6-cp37-cp37m-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (15.7 MB)\n",
      "Requirement already satisfied: python-dateutil in /home/dan/venv/pandas-bisect/lib/python3.7/site-packages (2.8.2)\n",
      "Collecting pytz\n",
      "  Using cached pytz-2023.3-py2.py3-none-any.whl (502 kB)\n",
      "Collecting scipy\n",
      "  Using cached scipy-1.7.3-cp37-cp37m-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (38.1 MB)\n",
      "Collecting versioneer[toml]\n",
      "  Using cached versioneer-0.29-py3-none-any.whl (46 kB)\n",
      "Requirement already satisfied: six>=1.5 in /home/dan/venv/pandas-bisect/lib/python3.7/site-packages (from python-dateutil) (1.16.0)\n",
      "Collecting tomli\n",
      "  Using cached tomli-2.0.1-py3-none-any.whl (12 kB)\n",
      "Installing collected packages: pytz, versioneer, tomli, numpy, cython, scipy\n",
      "Successfully installed cython-3.0.0 numpy-1.21.6 pytz-2023.3 scipy-1.7.3 tomli-2.0.1 versioneer-0.29\n",
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.0.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m23.2.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!cd pandas-dev && pip install cython numpy python-dateutil pytz scipy versioneer[toml]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f36a8184",
   "metadata": {
    "papermill": {
     "duration": 683.061945,
     "end_time": "2023-05-04T09:57:35.221747",
     "exception": false,
     "start_time": "2023-05-04T09:46:12.159802",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "_ = !cd pandas-dev && python setup.py build_ext -j 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "38139f10",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Compiling pandas/_libs/algos.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/arrays.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/groupby.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/hashing.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/hashtable.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/index.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/indexing.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/internals.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/interval.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/join.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/lib.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/missing.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/parsers.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/reduction.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/ops.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/ops_dispatch.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/properties.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/reshape.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/sparse.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/tslib.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/tslibs/base.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/tslibs/ccalendar.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/tslibs/dtypes.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/tslibs/conversion.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/tslibs/fields.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/tslibs/nattype.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/tslibs/np_datetime.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/tslibs/offsets.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/tslibs/parsing.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/tslibs/period.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/tslibs/strptime.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/tslibs/timedeltas.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/tslibs/timestamps.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/tslibs/timezones.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/tslibs/tzconversion.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/tslibs/vectorized.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/testing.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/window/aggregations.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/window/indexers.pyx because it changed.',\n",
       " 'Compiling pandas/_libs/writers.pyx because it changed.',\n",
       " 'Compiling pandas/io/sas/sas.pyx because it changed.',\n",
       " '[ 1/41] Cythonizing pandas/_libs/algos.pyx',\n",
       " '[ 2/41] Cythonizing pandas/_libs/arrays.pyx',\n",
       " '[ 3/41] Cythonizing pandas/_libs/groupby.pyx',\n",
       " '[ 4/41] Cythonizing pandas/_libs/hashing.pyx',\n",
       " '[ 5/41] Cythonizing pandas/_libs/hashtable.pyx',\n",
       " '[ 6/41] Cythonizing pandas/_libs/index.pyx',\n",
       " '[ 7/41] Cythonizing pandas/_libs/indexing.pyx',\n",
       " '[ 8/41] Cythonizing pandas/_libs/internals.pyx',\n",
       " '[ 9/41] Cythonizing pandas/_libs/interval.pyx',\n",
       " '[10/41] Cythonizing pandas/_libs/join.pyx',\n",
       " '[11/41] Cythonizing pandas/_libs/lib.pyx',\n",
       " '[12/41] Cythonizing pandas/_libs/missing.pyx',\n",
       " '[13/41] Cythonizing pandas/_libs/ops.pyx',\n",
       " '[14/41] Cythonizing pandas/_libs/ops_dispatch.pyx',\n",
       " '[15/41] Cythonizing pandas/_libs/parsers.pyx',\n",
       " '[16/41] Cythonizing pandas/_libs/properties.pyx',\n",
       " '[17/41] Cythonizing pandas/_libs/reduction.pyx',\n",
       " '[18/41] Cythonizing pandas/_libs/reshape.pyx',\n",
       " '[19/41] Cythonizing pandas/_libs/sparse.pyx',\n",
       " '[20/41] Cythonizing pandas/_libs/testing.pyx',\n",
       " '[21/41] Cythonizing pandas/_libs/tslib.pyx',\n",
       " '[22/41] Cythonizing pandas/_libs/tslibs/base.pyx',\n",
       " '[23/41] Cythonizing pandas/_libs/tslibs/ccalendar.pyx',\n",
       " '[24/41] Cythonizing pandas/_libs/tslibs/conversion.pyx',\n",
       " '[25/41] Cythonizing pandas/_libs/tslibs/dtypes.pyx',\n",
       " '[26/41] Cythonizing pandas/_libs/tslibs/fields.pyx',\n",
       " '[27/41] Cythonizing pandas/_libs/tslibs/nattype.pyx',\n",
       " '[28/41] Cythonizing pandas/_libs/tslibs/np_datetime.pyx',\n",
       " '[29/41] Cythonizing pandas/_libs/tslibs/offsets.pyx',\n",
       " '[30/41] Cythonizing pandas/_libs/tslibs/parsing.pyx',\n",
       " '[31/41] Cythonizing pandas/_libs/tslibs/period.pyx',\n",
       " '[32/41] Cythonizing pandas/_libs/tslibs/strptime.pyx',\n",
       " '[33/41] Cythonizing pandas/_libs/tslibs/timedeltas.pyx',\n",
       " '[34/41] Cythonizing pandas/_libs/tslibs/timestamps.pyx',\n",
       " '[35/41] Cythonizing pandas/_libs/tslibs/timezones.pyx',\n",
       " '[36/41] Cythonizing pandas/_libs/tslibs/tzconversion.pyx',\n",
       " '[37/41] Cythonizing pandas/_libs/tslibs/vectorized.pyx',\n",
       " '[38/41] Cythonizing pandas/_libs/window/aggregations.pyx',\n",
       " '[39/41] Cythonizing pandas/_libs/window/indexers.pyx',\n",
       " '[40/41] Cythonizing pandas/_libs/writers.pyx',\n",
       " '[41/41] Cythonizing pandas/io/sas/sas.pyx',\n",
       " \"warning: pandas/_libs/algos.pyx:153:42: Unknown type declaration 'bint' in annotation, ignoring\",\n",
       " \"warning: pandas/_libs/tslibs/nattype.pyx:92:0: Global name __nat_unpickle matched from within class scope in contradiction to to Python 'class private name' rules. This may change in a future release.\",\n",
       " \"warning: pandas/_libs/tslibs/nattype.pyx:92:0: Global name __nat_unpickle matched from within class scope in contradiction to to Python 'class private name' rules. This may change in a future release.\",\n",
       " \"warning: pandas/_libs/tslibs/offsets.pyx:4440:58: The keyword 'nogil' should appear at the end of the function signature line. Placing it before 'except' or 'noexcept' will be disallowed in a future version of Cython.\",\n",
       " \"warning: pandas/_libs/tslibs/parsing.pyx:905:29: Unknown type declaration 'bint' in annotation, ignoring\",\n",
       " '',\n",
       " 'Error compiling Cython file:',\n",
       " '------------------------------------------------------------',\n",
       " '...',\n",
       " '        if ((len(number) or len(frac)) and not len(unit)',\n",
       " '                and current_unit is None):',\n",
       " '            raise ValueError(\"no units specified\")',\n",
       " '',\n",
       " '        if len(frac) > 0 and len(frac) <= 3:',\n",
       " '            m = 10**(3 -len(frac)) * 1000 * 1000',\n",
       " '                                          ^',\n",
       " '------------------------------------------------------------',\n",
       " '',\n",
       " \"pandas/_libs/tslibs/timedeltas.pyx:611:42: Cannot assign type 'double' to 'int64_t'\",\n",
       " '',\n",
       " 'Error compiling Cython file:',\n",
       " '------------------------------------------------------------',\n",
       " '...',\n",
       " '            raise ValueError(\"no units specified\")',\n",
       " '',\n",
       " '        if len(frac) > 0 and len(frac) <= 3:',\n",
       " '            m = 10**(3 -len(frac)) * 1000 * 1000',\n",
       " '        elif len(frac) > 3 and len(frac) <= 6:',\n",
       " '            m = 10**(6 -len(frac)) * 1000',\n",
       " '                                   ^',\n",
       " '------------------------------------------------------------',\n",
       " '',\n",
       " \"pandas/_libs/tslibs/timedeltas.pyx:613:35: Cannot assign type 'double' to 'int64_t'\",\n",
       " '',\n",
       " 'Error compiling Cython file:',\n",
       " '------------------------------------------------------------',\n",
       " '...',\n",
       " '        if len(frac) > 0 and len(frac) <= 3:',\n",
       " '            m = 10**(3 -len(frac)) * 1000 * 1000',\n",
       " '        elif len(frac) > 3 and len(frac) <= 6:',\n",
       " '            m = 10**(6 -len(frac)) * 1000',\n",
       " '        elif len(frac) > 6 and len(frac) <= 9:',\n",
       " '            m = 10**(9 -len(frac))',\n",
       " '                  ^',\n",
       " '------------------------------------------------------------',\n",
       " '',\n",
       " \"pandas/_libs/tslibs/timedeltas.pyx:615:18: Cannot assign type 'double' to 'int64_t'\",\n",
       " 'Traceback (most recent call last):',\n",
       " '  File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1325, in cythonize_one_helper',\n",
       " '    return cythonize_one(*m)',\n",
       " '  File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1301, in cythonize_one',\n",
       " '    raise CompileError(None, pyx_file)',\n",
       " 'Cython.Compiler.Errors.CompileError: pandas/_libs/tslibs/timedeltas.pyx',\n",
       " '',\n",
       " 'Error compiling Cython file:',\n",
       " '------------------------------------------------------------',\n",
       " '...',\n",
       " '                if weighted == weighted:',\n",
       " '',\n",
       " '                    if is_observation or not ignore_na:',\n",
       " '                        if normalize:',\n",
       " '                            if use_deltas:',\n",
       " '                                old_wt *= old_wt_factor ** sub_deltas[i - 1]',\n",
       " '                                ^',\n",
       " '------------------------------------------------------------',\n",
       " '',\n",
       " \"pandas/_libs/window/aggregations.pyx:1795:32: Cannot assign type 'npy_float64 complex' to 'float64_t'\",\n",
       " 'Traceback (most recent call last):',\n",
       " '  File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1325, in cythonize_one_helper',\n",
       " '    return cythonize_one(*m)',\n",
       " '  File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1301, in cythonize_one',\n",
       " '    raise CompileError(None, pyx_file)',\n",
       " 'Cython.Compiler.Errors.CompileError: pandas/_libs/window/aggregations.pyx',\n",
       " \"warning: pandas/_libs/writers.pyx:123:50: Unknown type declaration 'Py_ssize_t' in annotation, ignoring\",\n",
       " \"warning: pandas/_libs/writers.pyx:123:50: Unknown type declaration 'Py_ssize_t' in annotation, ignoring\",\n",
       " \"warning: pandas/_libs/writers.pyx:123:50: Unknown type declaration 'Py_ssize_t' in annotation, ignoring\",\n",
       " 'multiprocessing.pool.RemoteTraceback: ',\n",
       " '\"\"\"',\n",
       " 'Traceback (most recent call last):',\n",
       " '  File \"/usr/lib/python3.7/multiprocessing/pool.py\", line 121, in worker',\n",
       " '    result = (True, func(*args, **kwds))',\n",
       " '  File \"/usr/lib/python3.7/multiprocessing/pool.py\", line 44, in mapstar',\n",
       " '    return list(map(*args))',\n",
       " '  File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1325, in cythonize_one_helper',\n",
       " '    return cythonize_one(*m)',\n",
       " '  File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1301, in cythonize_one',\n",
       " '    raise CompileError(None, pyx_file)',\n",
       " 'Cython.Compiler.Errors.CompileError: pandas/_libs/tslibs/timedeltas.pyx',\n",
       " '\"\"\"',\n",
       " '',\n",
       " 'The above exception was the direct cause of the following exception:',\n",
       " '',\n",
       " 'Traceback (most recent call last):',\n",
       " '  File \"setup.py\", line 664, in <module>',\n",
       " '    ext_modules=maybe_cythonize(extensions, compiler_directives=directives),',\n",
       " '  File \"setup.py\", line 424, in maybe_cythonize',\n",
       " '    return cythonize(extensions, *args, **kwargs)',\n",
       " '  File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1125, in cythonize',\n",
       " '    result.get(99999)  # seconds',\n",
       " '  File \"/usr/lib/python3.7/multiprocessing/pool.py\", line 657, in get',\n",
       " '    raise self._value',\n",
       " 'Cython.Compiler.Errors.CompileError: pandas/_libs/tslibs/timedeltas.pyx']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d7e6d110",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n"
     ]
    }
   ],
   "source": [
    "print(\"done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "24d09f10",
   "metadata": {
    "papermill": {
     "duration": 52.74671,
     "end_time": "2023-05-04T09:58:27.974238",
     "exception": false,
     "start_time": "2023-05-04T09:57:35.227528",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Obtaining file:///home/dan/notebooks/pandas-dev/pandas-dev\n",
      "  Preparing metadata (setup.py) ... \u001b[?25lerror\n",
      "  \u001b[1;31merror\u001b[0m: \u001b[1msubprocess-exited-with-error\u001b[0m\n",
      "  \n",
      "  \u001b[31m×\u001b[0m \u001b[32mpython setup.py egg_info\u001b[0m did not run successfully.\n",
      "  \u001b[31m│\u001b[0m exit code: \u001b[1;36m1\u001b[0m\n",
      "  \u001b[31m╰─>\u001b[0m \u001b[31m[100 lines of output]\u001b[0m\n",
      "  \u001b[31m   \u001b[0m Compiling pandas/_libs/tslibs/timedeltas.pyx because it changed.\n",
      "  \u001b[31m   \u001b[0m Compiling pandas/_libs/window/aggregations.pyx because it changed.\n",
      "  \u001b[31m   \u001b[0m [1/2] Cythonizing pandas/_libs/tslibs/timedeltas.pyx\n",
      "  \u001b[31m   \u001b[0m [2/2] Cythonizing pandas/_libs/window/aggregations.pyx\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m Error compiling Cython file:\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m ...\n",
      "  \u001b[31m   \u001b[0m         if ((len(number) or len(frac)) and not len(unit)\n",
      "  \u001b[31m   \u001b[0m                 and current_unit is None):\n",
      "  \u001b[31m   \u001b[0m             raise ValueError(\"no units specified\")\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m         if len(frac) > 0 and len(frac) <= 3:\n",
      "  \u001b[31m   \u001b[0m             m = 10**(3 -len(frac)) * 1000 * 1000\n",
      "  \u001b[31m   \u001b[0m                                           ^\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m pandas/_libs/tslibs/timedeltas.pyx:611:42: Cannot assign type 'double' to 'int64_t'\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m Error compiling Cython file:\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m ...\n",
      "  \u001b[31m   \u001b[0m             raise ValueError(\"no units specified\")\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m         if len(frac) > 0 and len(frac) <= 3:\n",
      "  \u001b[31m   \u001b[0m             m = 10**(3 -len(frac)) * 1000 * 1000\n",
      "  \u001b[31m   \u001b[0m         elif len(frac) > 3 and len(frac) <= 6:\n",
      "  \u001b[31m   \u001b[0m             m = 10**(6 -len(frac)) * 1000\n",
      "  \u001b[31m   \u001b[0m                                    ^\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m pandas/_libs/tslibs/timedeltas.pyx:613:35: Cannot assign type 'double' to 'int64_t'\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m Error compiling Cython file:\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m ...\n",
      "  \u001b[31m   \u001b[0m         if len(frac) > 0 and len(frac) <= 3:\n",
      "  \u001b[31m   \u001b[0m             m = 10**(3 -len(frac)) * 1000 * 1000\n",
      "  \u001b[31m   \u001b[0m         elif len(frac) > 3 and len(frac) <= 6:\n",
      "  \u001b[31m   \u001b[0m             m = 10**(6 -len(frac)) * 1000\n",
      "  \u001b[31m   \u001b[0m         elif len(frac) > 6 and len(frac) <= 9:\n",
      "  \u001b[31m   \u001b[0m             m = 10**(9 -len(frac))\n",
      "  \u001b[31m   \u001b[0m                   ^\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m pandas/_libs/tslibs/timedeltas.pyx:615:18: Cannot assign type 'double' to 'int64_t'\n",
      "  \u001b[31m   \u001b[0m Traceback (most recent call last):\n",
      "  \u001b[31m   \u001b[0m   File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1325, in cythonize_one_helper\n",
      "  \u001b[31m   \u001b[0m     return cythonize_one(*m)\n",
      "  \u001b[31m   \u001b[0m   File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1301, in cythonize_one\n",
      "  \u001b[31m   \u001b[0m     raise CompileError(None, pyx_file)\n",
      "  \u001b[31m   \u001b[0m Cython.Compiler.Errors.CompileError: pandas/_libs/tslibs/timedeltas.pyx\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m Error compiling Cython file:\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m ...\n",
      "  \u001b[31m   \u001b[0m                 if weighted == weighted:\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m                     if is_observation or not ignore_na:\n",
      "  \u001b[31m   \u001b[0m                         if normalize:\n",
      "  \u001b[31m   \u001b[0m                             if use_deltas:\n",
      "  \u001b[31m   \u001b[0m                                 old_wt *= old_wt_factor ** sub_deltas[i - 1]\n",
      "  \u001b[31m   \u001b[0m                                 ^\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m pandas/_libs/window/aggregations.pyx:1795:32: Cannot assign type 'npy_float64 complex' to 'float64_t'\n",
      "  \u001b[31m   \u001b[0m Traceback (most recent call last):\n",
      "  \u001b[31m   \u001b[0m   File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1325, in cythonize_one_helper\n",
      "  \u001b[31m   \u001b[0m     return cythonize_one(*m)\n",
      "  \u001b[31m   \u001b[0m   File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1301, in cythonize_one\n",
      "  \u001b[31m   \u001b[0m     raise CompileError(None, pyx_file)\n",
      "  \u001b[31m   \u001b[0m Cython.Compiler.Errors.CompileError: pandas/_libs/window/aggregations.pyx\n",
      "  \u001b[31m   \u001b[0m multiprocessing.pool.RemoteTraceback:\n",
      "  \u001b[31m   \u001b[0m \"\"\"\n",
      "  \u001b[31m   \u001b[0m Traceback (most recent call last):\n",
      "  \u001b[31m   \u001b[0m   File \"/usr/lib/python3.7/multiprocessing/pool.py\", line 121, in worker\n",
      "  \u001b[31m   \u001b[0m     result = (True, func(*args, **kwds))\n",
      "  \u001b[31m   \u001b[0m   File \"/usr/lib/python3.7/multiprocessing/pool.py\", line 44, in mapstar\n",
      "  \u001b[31m   \u001b[0m     return list(map(*args))\n",
      "  \u001b[31m   \u001b[0m   File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1325, in cythonize_one_helper\n",
      "  \u001b[31m   \u001b[0m     return cythonize_one(*m)\n",
      "  \u001b[31m   \u001b[0m   File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1301, in cythonize_one\n",
      "  \u001b[31m   \u001b[0m     raise CompileError(None, pyx_file)\n",
      "  \u001b[31m   \u001b[0m Cython.Compiler.Errors.CompileError: pandas/_libs/tslibs/timedeltas.pyx\n",
      "  \u001b[31m   \u001b[0m \"\"\"\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m The above exception was the direct cause of the following exception:\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m Traceback (most recent call last):\n",
      "  \u001b[31m   \u001b[0m   File \"<string>\", line 36, in <module>\n",
      "  \u001b[31m   \u001b[0m   File \"<pip-setuptools-caller>\", line 34, in <module>\n",
      "  \u001b[31m   \u001b[0m   File \"/home/dan/notebooks/pandas-dev/pandas-dev/setup.py\", line 664, in <module>\n",
      "  \u001b[31m   \u001b[0m     ext_modules=maybe_cythonize(extensions, compiler_directives=directives),\n",
      "  \u001b[31m   \u001b[0m   File \"/home/dan/notebooks/pandas-dev/pandas-dev/setup.py\", line 424, in maybe_cythonize\n",
      "  \u001b[31m   \u001b[0m     return cythonize(extensions, *args, **kwargs)\n",
      "  \u001b[31m   \u001b[0m   File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1125, in cythonize\n",
      "  \u001b[31m   \u001b[0m     result.get(99999)  # seconds\n",
      "  \u001b[31m   \u001b[0m   File \"/usr/lib/python3.7/multiprocessing/pool.py\", line 657, in get\n",
      "  \u001b[31m   \u001b[0m     raise self._value\n",
      "  \u001b[31m   \u001b[0m Cython.Compiler.Errors.CompileError: pandas/_libs/tslibs/timedeltas.pyx\n",
      "  \u001b[31m   \u001b[0m \u001b[31m[end of output]\u001b[0m\n",
      "  \n",
      "  \u001b[1;35mnote\u001b[0m: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "\u001b[1;31merror\u001b[0m: \u001b[1mmetadata-generation-failed\u001b[0m\n",
      "\n",
      "\u001b[31m×\u001b[0m Encountered error while generating package metadata.\n",
      "\u001b[31m╰─>\u001b[0m See above for output.\n",
      "\n",
      "\u001b[1;35mnote\u001b[0m: This is an issue with the package mentioned above, not pip.\n",
      "\u001b[1;36mhint\u001b[0m: See above for details.\n",
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.0.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m23.2.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n",
      "\u001b[?25h"
     ]
    }
   ],
   "source": [
    "!cd pandas-dev && pip install -e . --no-build-isolation --no-use-pep517"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "40e53118",
   "metadata": {
    "papermill": {
     "duration": 2.847816,
     "end_time": "2023-05-04T09:58:30.827913",
     "exception": false,
     "start_time": "2023-05-04T09:58:27.980097",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\r\n",
      "  File \"<string>\", line 1, in <module>\r\n",
      "  File \"/home/dan/notebooks/pandas-dev/pandas-dev/pandas/__init__.py\", line 22, in <module>\r\n",
      "    from pandas.compat import is_numpy_dev as _is_numpy_dev  # pyright: ignore # noqa:F401\r\n",
      "  File \"/home/dan/notebooks/pandas-dev/pandas-dev/pandas/compat/__init__.py\", line 17, in <module>\r\n",
      "    from pandas._typing import F\r\n",
      "  File \"/home/dan/notebooks/pandas-dev/pandas-dev/pandas/_typing.py\", line 9, in <module>\r\n",
      "    from typing import (\r\n",
      "ImportError: cannot import name 'Literal' from 'typing' (/usr/lib/python3.7/typing.py)\r\n"
     ]
    }
   ],
   "source": [
    "!cd pandas-dev && python -c 'import pandas; print(pandas.__version__)'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "449d9809",
   "metadata": {
    "papermill": {
     "duration": 114.074437,
     "end_time": "2023-05-04T10:00:24.909645",
     "exception": false,
     "start_time": "2023-05-04T09:58:30.835208",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "has versioneer:  False\n",
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.0.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m23.2.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n",
      "Found existing installation: versioneer 0.29\n",
      "Uninstalling versioneer-0.29:\n",
      "  Successfully uninstalled versioneer-0.29\n",
      "Compiling pandas/_libs/tslibs/timedeltas.pyx because it changed.\n",
      "Compiling pandas/_libs/window/aggregations.pyx because it changed.\n",
      "[1/2] Cythonizing pandas/_libs/tslibs/timedeltas.pyx\n",
      "\n",
      "Error compiling Cython file:\n",
      "------------------------------------------------------------\n",
      "...\n",
      "        if ((len(number) or len(frac)) and not len(unit)\n",
      "                and current_unit is None):\n",
      "            raise ValueError(\"no units specified\")\n",
      "\n",
      "        if len(frac) > 0 and len(frac) <= 3:\n",
      "            m = 10**(3 -len(frac)) * 1000 * 1000\n",
      "                                          ^\n",
      "------------------------------------------------------------\n",
      "\n",
      "pandas/_libs/tslibs/timedeltas.pyx:611:42: Cannot assign type 'double' to 'int64_t'\n",
      "\n",
      "Error compiling Cython file:\n",
      "------------------------------------------------------------\n",
      "...\n",
      "            raise ValueError(\"no units specified\")\n",
      "\n",
      "        if len(frac) > 0 and len(frac) <= 3:\n",
      "            m = 10**(3 -len(frac)) * 1000 * 1000\n",
      "        elif len(frac) > 3 and len(frac) <= 6:\n",
      "            m = 10**(6 -len(frac)) * 1000\n",
      "                                   ^\n",
      "------------------------------------------------------------\n",
      "\n",
      "pandas/_libs/tslibs/timedeltas.pyx:613:35: Cannot assign type 'double' to 'int64_t'\n",
      "\n",
      "Error compiling Cython file:\n",
      "------------------------------------------------------------\n",
      "...\n",
      "        if len(frac) > 0 and len(frac) <= 3:\n",
      "            m = 10**(3 -len(frac)) * 1000 * 1000\n",
      "        elif len(frac) > 3 and len(frac) <= 6:\n",
      "            m = 10**(6 -len(frac)) * 1000\n",
      "        elif len(frac) > 6 and len(frac) <= 9:\n",
      "            m = 10**(9 -len(frac))\n",
      "                  ^\n",
      "------------------------------------------------------------\n",
      "\n",
      "pandas/_libs/tslibs/timedeltas.pyx:615:18: Cannot assign type 'double' to 'int64_t'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1325, in cythonize_one_helper\n",
      "    return cythonize_one(*m)\n",
      "  File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1301, in cythonize_one\n",
      "    raise CompileError(None, pyx_file)\n",
      "Cython.Compiler.Errors.CompileError: pandas/_libs/tslibs/timedeltas.pyx\n",
      "[2/2] Cythonizing pandas/_libs/window/aggregations.pyx\n",
      "\n",
      "Error compiling Cython file:\n",
      "------------------------------------------------------------\n",
      "...\n",
      "                if weighted == weighted:\n",
      "\n",
      "                    if is_observation or not ignore_na:\n",
      "                        if normalize:\n",
      "                            if use_deltas:\n",
      "                                old_wt *= old_wt_factor ** sub_deltas[i - 1]\n",
      "                                ^\n",
      "------------------------------------------------------------\n",
      "\n",
      "pandas/_libs/window/aggregations.pyx:1795:32: Cannot assign type 'npy_float64 complex' to 'float64_t'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1325, in cythonize_one_helper\n",
      "    return cythonize_one(*m)\n",
      "  File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1301, in cythonize_one\n",
      "    raise CompileError(None, pyx_file)\n",
      "Cython.Compiler.Errors.CompileError: pandas/_libs/window/aggregations.pyx\n",
      "multiprocessing.pool.RemoteTraceback: \n",
      "\"\"\"\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/lib/python3.7/multiprocessing/pool.py\", line 121, in worker\n",
      "    result = (True, func(*args, **kwds))\n",
      "  File \"/usr/lib/python3.7/multiprocessing/pool.py\", line 44, in mapstar\n",
      "    return list(map(*args))\n",
      "  File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1325, in cythonize_one_helper\n",
      "    return cythonize_one(*m)\n",
      "  File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1301, in cythonize_one\n",
      "    raise CompileError(None, pyx_file)\n",
      "Cython.Compiler.Errors.CompileError: pandas/_libs/tslibs/timedeltas.pyx\n",
      "\"\"\"\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"setup.py\", line 664, in <module>\n",
      "    ext_modules=maybe_cythonize(extensions, compiler_directives=directives),\n",
      "  File \"setup.py\", line 424, in maybe_cythonize\n",
      "    return cythonize(extensions, *args, **kwargs)\n",
      "  File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1125, in cythonize\n",
      "    result.get(99999)  # seconds\n",
      "  File \"/usr/lib/python3.7/multiprocessing/pool.py\", line 657, in get\n",
      "    raise self._value\n",
      "Cython.Compiler.Errors.CompileError: pandas/_libs/tslibs/timedeltas.pyx\n",
      "Obtaining file:///home/dan/notebooks/pandas-dev/pandas-dev\n",
      "  Preparing metadata (setup.py) ... \u001b[?25lerror\n",
      "  \u001b[1;31merror\u001b[0m: \u001b[1msubprocess-exited-with-error\u001b[0m\n",
      "  \n",
      "  \u001b[31m×\u001b[0m \u001b[32mpython setup.py egg_info\u001b[0m did not run successfully.\n",
      "  \u001b[31m│\u001b[0m exit code: \u001b[1;36m1\u001b[0m\n",
      "  \u001b[31m╰─>\u001b[0m \u001b[31m[100 lines of output]\u001b[0m\n",
      "  \u001b[31m   \u001b[0m Compiling pandas/_libs/tslibs/timedeltas.pyx because it changed.\n",
      "  \u001b[31m   \u001b[0m Compiling pandas/_libs/window/aggregations.pyx because it changed.\n",
      "  \u001b[31m   \u001b[0m [1/2] Cythonizing pandas/_libs/tslibs/timedeltas.pyx\n",
      "  \u001b[31m   \u001b[0m [2/2] Cythonizing pandas/_libs/window/aggregations.pyx\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m Error compiling Cython file:\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m ...\n",
      "  \u001b[31m   \u001b[0m         if ((len(number) or len(frac)) and not len(unit)\n",
      "  \u001b[31m   \u001b[0m                 and current_unit is None):\n",
      "  \u001b[31m   \u001b[0m             raise ValueError(\"no units specified\")\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m         if len(frac) > 0 and len(frac) <= 3:\n",
      "  \u001b[31m   \u001b[0m             m = 10**(3 -len(frac)) * 1000 * 1000\n",
      "  \u001b[31m   \u001b[0m                                           ^\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m pandas/_libs/tslibs/timedeltas.pyx:611:42: Cannot assign type 'double' to 'int64_t'\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m Error compiling Cython file:\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m ...\n",
      "  \u001b[31m   \u001b[0m             raise ValueError(\"no units specified\")\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m         if len(frac) > 0 and len(frac) <= 3:\n",
      "  \u001b[31m   \u001b[0m             m = 10**(3 -len(frac)) * 1000 * 1000\n",
      "  \u001b[31m   \u001b[0m         elif len(frac) > 3 and len(frac) <= 6:\n",
      "  \u001b[31m   \u001b[0m             m = 10**(6 -len(frac)) * 1000\n",
      "  \u001b[31m   \u001b[0m                                    ^\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m pandas/_libs/tslibs/timedeltas.pyx:613:35: Cannot assign type 'double' to 'int64_t'\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m Error compiling Cython file:\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m ...\n",
      "  \u001b[31m   \u001b[0m         if len(frac) > 0 and len(frac) <= 3:\n",
      "  \u001b[31m   \u001b[0m             m = 10**(3 -len(frac)) * 1000 * 1000\n",
      "  \u001b[31m   \u001b[0m         elif len(frac) > 3 and len(frac) <= 6:\n",
      "  \u001b[31m   \u001b[0m             m = 10**(6 -len(frac)) * 1000\n",
      "  \u001b[31m   \u001b[0m         elif len(frac) > 6 and len(frac) <= 9:\n",
      "  \u001b[31m   \u001b[0m             m = 10**(9 -len(frac))\n",
      "  \u001b[31m   \u001b[0m                   ^\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m pandas/_libs/tslibs/timedeltas.pyx:615:18: Cannot assign type 'double' to 'int64_t'\n",
      "  \u001b[31m   \u001b[0m Traceback (most recent call last):\n",
      "  \u001b[31m   \u001b[0m   File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1325, in cythonize_one_helper\n",
      "  \u001b[31m   \u001b[0m     return cythonize_one(*m)\n",
      "  \u001b[31m   \u001b[0m   File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1301, in cythonize_one\n",
      "  \u001b[31m   \u001b[0m     raise CompileError(None, pyx_file)\n",
      "  \u001b[31m   \u001b[0m Cython.Compiler.Errors.CompileError: pandas/_libs/tslibs/timedeltas.pyx\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m Error compiling Cython file:\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m ...\n",
      "  \u001b[31m   \u001b[0m                 if weighted == weighted:\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m                     if is_observation or not ignore_na:\n",
      "  \u001b[31m   \u001b[0m                         if normalize:\n",
      "  \u001b[31m   \u001b[0m                             if use_deltas:\n",
      "  \u001b[31m   \u001b[0m                                 old_wt *= old_wt_factor ** sub_deltas[i - 1]\n",
      "  \u001b[31m   \u001b[0m                                 ^\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m pandas/_libs/window/aggregations.pyx:1795:32: Cannot assign type 'npy_float64 complex' to 'float64_t'\n",
      "  \u001b[31m   \u001b[0m Traceback (most recent call last):\n",
      "  \u001b[31m   \u001b[0m   File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1325, in cythonize_one_helper\n",
      "  \u001b[31m   \u001b[0m     return cythonize_one(*m)\n",
      "  \u001b[31m   \u001b[0m   File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1301, in cythonize_one\n",
      "  \u001b[31m   \u001b[0m     raise CompileError(None, pyx_file)\n",
      "  \u001b[31m   \u001b[0m Cython.Compiler.Errors.CompileError: pandas/_libs/window/aggregations.pyx\n",
      "  \u001b[31m   \u001b[0m multiprocessing.pool.RemoteTraceback:\n",
      "  \u001b[31m   \u001b[0m \"\"\"\n",
      "  \u001b[31m   \u001b[0m Traceback (most recent call last):\n",
      "  \u001b[31m   \u001b[0m   File \"/usr/lib/python3.7/multiprocessing/pool.py\", line 121, in worker\n",
      "  \u001b[31m   \u001b[0m     result = (True, func(*args, **kwds))\n",
      "  \u001b[31m   \u001b[0m   File \"/usr/lib/python3.7/multiprocessing/pool.py\", line 44, in mapstar\n",
      "  \u001b[31m   \u001b[0m     return list(map(*args))\n",
      "  \u001b[31m   \u001b[0m   File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1325, in cythonize_one_helper\n",
      "  \u001b[31m   \u001b[0m     return cythonize_one(*m)\n",
      "  \u001b[31m   \u001b[0m   File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1301, in cythonize_one\n",
      "  \u001b[31m   \u001b[0m     raise CompileError(None, pyx_file)\n",
      "  \u001b[31m   \u001b[0m Cython.Compiler.Errors.CompileError: pandas/_libs/tslibs/timedeltas.pyx\n",
      "  \u001b[31m   \u001b[0m \"\"\"\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m The above exception was the direct cause of the following exception:\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m Traceback (most recent call last):\n",
      "  \u001b[31m   \u001b[0m   File \"<string>\", line 36, in <module>\n",
      "  \u001b[31m   \u001b[0m   File \"<pip-setuptools-caller>\", line 34, in <module>\n",
      "  \u001b[31m   \u001b[0m   File \"/home/dan/notebooks/pandas-dev/pandas-dev/setup.py\", line 664, in <module>\n",
      "  \u001b[31m   \u001b[0m     ext_modules=maybe_cythonize(extensions, compiler_directives=directives),\n",
      "  \u001b[31m   \u001b[0m   File \"/home/dan/notebooks/pandas-dev/pandas-dev/setup.py\", line 424, in maybe_cythonize\n",
      "  \u001b[31m   \u001b[0m     return cythonize(extensions, *args, **kwargs)\n",
      "  \u001b[31m   \u001b[0m   File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1125, in cythonize\n",
      "  \u001b[31m   \u001b[0m     result.get(99999)  # seconds\n",
      "  \u001b[31m   \u001b[0m   File \"/usr/lib/python3.7/multiprocessing/pool.py\", line 657, in get\n",
      "  \u001b[31m   \u001b[0m     raise self._value\n",
      "  \u001b[31m   \u001b[0m Cython.Compiler.Errors.CompileError: pandas/_libs/tslibs/timedeltas.pyx\n",
      "  \u001b[31m   \u001b[0m \u001b[31m[end of output]\u001b[0m\n",
      "  \n",
      "  \u001b[1;35mnote\u001b[0m: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "\u001b[1;31merror\u001b[0m: \u001b[1mmetadata-generation-failed\u001b[0m\n",
      "\n",
      "\u001b[31m×\u001b[0m Encountered error while generating package metadata.\n",
      "\u001b[31m╰─>\u001b[0m See above for output.\n",
      "\n",
      "\u001b[1;35mnote\u001b[0m: This is an issue with the package mentioned above, not pip.\n",
      "\u001b[1;36mhint\u001b[0m: See above for details.\n",
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.0.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m23.2.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[?25hTraceback (most recent call last):\r\n",
      "  File \"t.py\", line 2, in <module>\r\n",
      "    import pandas as pd\r\n",
      "  File \"/home/dan/notebooks/pandas-dev/pandas-dev/pandas/__init__.py\", line 22, in <module>\r\n",
      "    from pandas.compat import is_numpy_dev as _is_numpy_dev  # pyright: ignore # noqa:F401\r\n",
      "  File \"/home/dan/notebooks/pandas-dev/pandas-dev/pandas/compat/__init__.py\", line 17, in <module>\r\n",
      "    from pandas._typing import F\r\n",
      "  File \"/home/dan/notebooks/pandas-dev/pandas-dev/pandas/_typing.py\", line 9, in <module>\r\n",
      "    from typing import (\r\n",
      "ImportError: cannot import name 'Literal' from 'typing' (/usr/lib/python3.7/typing.py)\r\n"
     ]
    }
   ],
   "source": [
    "!cd pandas-dev && bash bisect.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "56dbbd3f",
   "metadata": {
    "papermill": {
     "duration": 1.454977,
     "end_time": "2023-05-04T10:00:26.372064",
     "exception": false,
     "start_time": "2023-05-04T10:00:24.917087",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Previous HEAD position was 2e218d1098 RLS: 1.5.3\r\n",
      "HEAD is now at 1ce1c3c1ef RLS: 1.0.4\r\n"
     ]
    }
   ],
   "source": [
    "!cd pandas-dev && bash checkout_good.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "024ae909",
   "metadata": {
    "papermill": {
     "duration": 699.716278,
     "end_time": "2023-05-04T10:12:06.095722",
     "exception": false,
     "start_time": "2023-05-04T10:00:26.379444",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "has versioneer:  False\n",
      "Collecting git+https://github.com/pandas-dev/pydata-sphinx-theme.git@master\n",
      "  Cloning https://github.com/pandas-dev/pydata-sphinx-theme.git (to revision master) to /tmp/pip-req-build-muqq300t\n",
      "  Running command git clone --filter=blob:none --quiet https://github.com/pandas-dev/pydata-sphinx-theme.git /tmp/pip-req-build-muqq300t\n",
      "\u001b[33m  WARNING: Did not find branch or tag 'master', assuming revision or ref.\u001b[0m\u001b[33m\n",
      "\u001b[0m  Running command git checkout -q master\n",
      "  error: pathspec 'master' did not match any file(s) known to git\n",
      "  \u001b[1;31merror\u001b[0m: \u001b[1msubprocess-exited-with-error\u001b[0m\n",
      "  \n",
      "  \u001b[31m×\u001b[0m \u001b[32mgit checkout -q master\u001b[0m did not run successfully.\n",
      "  \u001b[31m│\u001b[0m exit code: \u001b[1;36m1\u001b[0m\n",
      "  \u001b[31m╰─>\u001b[0m See above for output.\n",
      "  \n",
      "  \u001b[1;35mnote\u001b[0m: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "\u001b[1;31merror\u001b[0m: \u001b[1msubprocess-exited-with-error\u001b[0m\n",
      "\n",
      "\u001b[31m×\u001b[0m \u001b[32mgit checkout -q master\u001b[0m did not run successfully.\n",
      "\u001b[31m│\u001b[0m exit code: \u001b[1;36m1\u001b[0m\n",
      "\u001b[31m╰─>\u001b[0m See above for output.\n",
      "\n",
      "\u001b[1;35mnote\u001b[0m: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.0.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m23.2.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n",
      "\u001b[33mWARNING: Skipping versioneer as it is not installed.\u001b[0m\u001b[33m\n",
      "\u001b[0mCompiling pandas/_libs/algos.pyx because it changed.\n",
      "Compiling pandas/_libs/groupby.pyx because it changed.\n",
      "Compiling pandas/_libs/hashing.pyx because it changed.\n",
      "Compiling pandas/_libs/hashtable.pyx because it changed.\n",
      "Compiling pandas/_libs/index.pyx because it changed.\n",
      "Compiling pandas/_libs/indexing.pyx because it changed.\n",
      "Compiling pandas/_libs/internals.pyx because it changed.\n",
      "Compiling pandas/_libs/interval.pyx because it changed.\n",
      "Compiling pandas/_libs/join.pyx because it changed.\n",
      "Compiling pandas/_libs/lib.pyx because it changed.\n",
      "Compiling pandas/_libs/missing.pyx because it changed.\n",
      "Compiling pandas/_libs/parsers.pyx because it changed.\n",
      "Compiling pandas/_libs/reduction.pyx because it changed.\n",
      "Compiling pandas/_libs/ops.pyx because it changed.\n",
      "Compiling pandas/_libs/ops_dispatch.pyx because it changed.\n",
      "Compiling pandas/_libs/properties.pyx because it changed.\n",
      "Compiling pandas/_libs/reshape.pyx because it changed.\n",
      "Compiling pandas/_libs/sparse.pyx because it changed.\n",
      "Compiling pandas/_libs/tslib.pyx because it changed.\n",
      "Compiling pandas/_libs/tslibs/c_timestamp.pyx because it changed.\n",
      "Compiling pandas/_libs/tslibs/ccalendar.pyx because it changed.\n",
      "Compiling pandas/_libs/tslibs/conversion.pyx because it changed.\n",
      "Compiling pandas/_libs/tslibs/fields.pyx because it changed.\n",
      "Compiling pandas/_libs/tslibs/frequencies.pyx because it changed.\n",
      "Compiling pandas/_libs/tslibs/nattype.pyx because it changed.\n",
      "Compiling pandas/_libs/tslibs/np_datetime.pyx because it changed.\n",
      "Compiling pandas/_libs/tslibs/offsets.pyx because it changed.\n",
      "Compiling pandas/_libs/tslibs/parsing.pyx because it changed.\n",
      "Compiling pandas/_libs/tslibs/period.pyx because it changed.\n",
      "Compiling pandas/_libs/tslibs/resolution.pyx because it changed.\n",
      "Compiling pandas/_libs/tslibs/strptime.pyx because it changed.\n",
      "Compiling pandas/_libs/tslibs/timedeltas.pyx because it changed.\n",
      "Compiling pandas/_libs/tslibs/timestamps.pyx because it changed.\n",
      "Compiling pandas/_libs/tslibs/timezones.pyx because it changed.\n",
      "Compiling pandas/_libs/tslibs/tzconversion.pyx because it changed.\n",
      "Compiling pandas/_libs/testing.pyx because it changed.\n",
      "Compiling pandas/_libs/window/aggregations.pyx because it changed.\n",
      "Compiling pandas/_libs/window/indexers.pyx because it changed.\n",
      "Compiling pandas/_libs/writers.pyx because it changed.\n",
      "Compiling pandas/io/sas/sas.pyx because it changed.\n",
      "[ 1/40] Cythonizing pandas/_libs/algos.pyx\n",
      "warning: pandas/_libs/algos_take_helper.pxi:51:4: The 'IF' statement is deprecated and will be removed in a future Cython version. Consider using runtime conditions or C macros instead. See https://github.com/cython/cython/issues/4310\n",
      "warning: pandas/_libs/algos_take_helper.pxi:178:4: The 'IF' statement is deprecated and will be removed in a future Cython version. Consider using runtime conditions or C macros instead. See https://github.com/cython/cython/issues/4310\n",
      "warning: pandas/_libs/algos_take_helper.pxi:305:4: The 'IF' statement is deprecated and will be removed in a future Cython version. Consider using runtime conditions or C macros instead. See https://github.com/cython/cython/issues/4310\n",
      "warning: pandas/_libs/algos_take_helper.pxi:432:4: The 'IF' statement is deprecated and will be removed in a future Cython version. Consider using runtime conditions or C macros instead. See https://github.com/cython/cython/issues/4310\n",
      "warning: pandas/_libs/algos_take_helper.pxi:559:4: The 'IF' statement is deprecated and will be removed in a future Cython version. Consider using runtime conditions or C macros instead. See https://github.com/cython/cython/issues/4310\n",
      "warning: pandas/_libs/algos_take_helper.pxi:686:4: The 'IF' statement is deprecated and will be removed in a future Cython version. Consider using runtime conditions or C macros instead. See https://github.com/cython/cython/issues/4310\n",
      "warning: pandas/_libs/algos_take_helper.pxi:813:4: The 'IF' statement is deprecated and will be removed in a future Cython version. Consider using runtime conditions or C macros instead. See https://github.com/cython/cython/issues/4310\n",
      "warning: pandas/_libs/algos_take_helper.pxi:940:4: The 'IF' statement is deprecated and will be removed in a future Cython version. Consider using runtime conditions or C macros instead. See https://github.com/cython/cython/issues/4310\n",
      "warning: pandas/_libs/algos_take_helper.pxi:1067:4: The 'IF' statement is deprecated and will be removed in a future Cython version. Consider using runtime conditions or C macros instead. See https://github.com/cython/cython/issues/4310\n",
      "warning: pandas/_libs/algos_take_helper.pxi:1194:4: The 'IF' statement is deprecated and will be removed in a future Cython version. Consider using runtime conditions or C macros instead. See https://github.com/cython/cython/issues/4310\n",
      "warning: pandas/_libs/algos_take_helper.pxi:1321:4: The 'IF' statement is deprecated and will be removed in a future Cython version. Consider using runtime conditions or C macros instead. See https://github.com/cython/cython/issues/4310\n",
      "warning: pandas/_libs/algos_take_helper.pxi:1448:4: The 'IF' statement is deprecated and will be removed in a future Cython version. Consider using runtime conditions or C macros instead. See https://github.com/cython/cython/issues/4310\n",
      "warning: pandas/_libs/algos_take_helper.pxi:1575:4: The 'IF' statement is deprecated and will be removed in a future Cython version. Consider using runtime conditions or C macros instead. See https://github.com/cython/cython/issues/4310\n",
      "warning: pandas/_libs/algos_take_helper.pxi:1702:4: The 'IF' statement is deprecated and will be removed in a future Cython version. Consider using runtime conditions or C macros instead. See https://github.com/cython/cython/issues/4310\n",
      "warning: pandas/_libs/algos_take_helper.pxi:1829:4: The 'IF' statement is deprecated and will be removed in a future Cython version. Consider using runtime conditions or C macros instead. See https://github.com/cython/cython/issues/4310\n",
      "warning: pandas/_libs/algos_take_helper.pxi:1956:4: The 'IF' statement is deprecated and will be removed in a future Cython version. Consider using runtime conditions or C macros instead. See https://github.com/cython/cython/issues/4310\n",
      "warning: pandas/_libs/algos_take_helper.pxi:2083:4: The 'IF' statement is deprecated and will be removed in a future Cython version. Consider using runtime conditions or C macros instead. See https://github.com/cython/cython/issues/4310\n",
      "warning: pandas/_libs/algos_take_helper.pxi:2210:4: The 'IF' statement is deprecated and will be removed in a future Cython version. Consider using runtime conditions or C macros instead. See https://github.com/cython/cython/issues/4310\n",
      "warning: pandas/_libs/algos_take_helper.pxi:2337:4: The 'IF' statement is deprecated and will be removed in a future Cython version. Consider using runtime conditions or C macros instead. See https://github.com/cython/cython/issues/4310\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "warning: pandas/_libs/algos.pyx:117:42: Unknown type declaration 'bint' in annotation, ignoring\n",
      "[ 2/40] Cythonizing pandas/_libs/groupby.pyx\n",
      "warning: pandas/_libs/groupby.pyx:1104:26: Unreachable code\n",
      "[ 3/40] Cythonizing pandas/_libs/hashing.pyx\n",
      "warning: pandas/_libs/hashing.pyx:13:0: The 'DEF' statement is deprecated and will be removed in a future Cython version. Consider using global variables, constants, and in-place literals instead. See https://github.com/cython/cython/issues/4310\n",
      "warning: pandas/_libs/hashing.pyx:14:0: The 'DEF' statement is deprecated and will be removed in a future Cython version. Consider using global variables, constants, and in-place literals instead. See https://github.com/cython/cython/issues/4310\n",
      "[ 4/40] Cythonizing pandas/_libs/hashtable.pyx\n",
      "[ 5/40] Cythonizing pandas/_libs/index.pyx\n",
      "[ 6/40] Cythonizing pandas/_libs/indexing.pyx\n",
      "[ 7/40] Cythonizing pandas/_libs/internals.pyx\n",
      "[ 8/40] Cythonizing pandas/_libs/interval.pyx\n",
      "[ 9/40] Cythonizing pandas/_libs/join.pyx\n",
      "[10/40] Cythonizing pandas/_libs/lib.pyx\n",
      "[11/40] Cythonizing pandas/_libs/missing.pyx\n",
      "[12/40] Cythonizing pandas/_libs/ops.pyx\n",
      "[13/40] Cythonizing pandas/_libs/ops_dispatch.pyx\n",
      "[14/40] Cythonizing pandas/_libs/parsers.pyx\n",
      "[15/40] Cythonizing pandas/_libs/properties.pyx\n",
      "[16/40] Cythonizing pandas/_libs/reduction.pyx\n",
      "\n",
      "Error compiling Cython file:\n",
      "------------------------------------------------------------\n",
      "...\n",
      "            object cached_typ = None\n",
      "\n",
      "        arr = self.arr\n",
      "        chunk = self.dummy\n",
      "        dummy_buf = chunk.data\n",
      "        chunk.data = arr.data\n",
      "             ^\n",
      "------------------------------------------------------------\n",
      "\n",
      "pandas/_libs/reduction.pyx:101:13: Assignment to a read-only property\n",
      "\n",
      "Error compiling Cython file:\n",
      "------------------------------------------------------------\n",
      "...\n",
      "                    # On the first pass, we check the output shape to see\n",
      "                    #  if this looks like a reduction.\n",
      "                    _check_result_array(res, len(self.dummy))\n",
      "\n",
      "                PyArray_SETITEM(result, PyArray_ITER_DATA(it), res)\n",
      "                chunk.data = chunk.data + self.increment\n",
      "                     ^\n",
      "------------------------------------------------------------\n",
      "\n",
      "pandas/_libs/reduction.pyx:140:21: Assignment to a read-only property\n",
      "\n",
      "Error compiling Cython file:\n",
      "------------------------------------------------------------\n",
      "...\n",
      "                PyArray_SETITEM(result, PyArray_ITER_DATA(it), res)\n",
      "                chunk.data = chunk.data + self.increment\n",
      "                PyArray_ITER_NEXT(it)\n",
      "        finally:\n",
      "            # so we don't free the wrong memory\n",
      "            chunk.data = dummy_buf\n",
      "                 ^\n",
      "------------------------------------------------------------\n",
      "\n",
      "pandas/_libs/reduction.pyx:144:17: Assignment to a read-only property\n",
      "\n",
      "Error compiling Cython file:\n",
      "------------------------------------------------------------\n",
      "...\n",
      "                PyArray_SETITEM(result, PyArray_ITER_DATA(it), res)\n",
      "                chunk.data = chunk.data + self.increment\n",
      "                PyArray_ITER_NEXT(it)\n",
      "        finally:\n",
      "            # so we don't free the wrong memory\n",
      "            chunk.data = dummy_buf\n",
      "                 ^\n",
      "------------------------------------------------------------\n",
      "\n",
      "pandas/_libs/reduction.pyx:144:17: Assignment to a read-only property\n",
      "\n",
      "Error compiling Cython file:\n",
      "------------------------------------------------------------\n",
      "...\n",
      "\n",
      "        self.orig_data = self.buf.data\n",
      "        self.orig_len = self.buf.shape[0]\n",
      "        self.orig_stride = self.buf.strides[0]\n",
      "\n",
      "        self.buf.data = self.values.data\n",
      "                ^\n",
      "------------------------------------------------------------\n",
      "\n",
      "pandas/_libs/reduction.pyx:437:16: Assignment to a read-only property\n",
      "\n",
      "Error compiling Cython file:\n",
      "------------------------------------------------------------\n",
      "...\n",
      "\n",
      "        self.buf.data = self.values.data\n",
      "        self.buf.strides[0] = self.stride\n",
      "\n",
      "    cdef advance(self, Py_ssize_t k):\n",
      "        self.buf.data = <char*>self.buf.data + self.stride * k\n",
      "                ^\n",
      "------------------------------------------------------------\n",
      "\n",
      "pandas/_libs/reduction.pyx:441:16: Assignment to a read-only property\n",
      "\n",
      "Error compiling Cython file:\n",
      "------------------------------------------------------------\n",
      "...\n",
      "\n",
      "    cdef move(self, int start, int end):\n",
      "        \"\"\"\n",
      "        For slicing\n",
      "        \"\"\"\n",
      "        self.buf.data = self.values.data + self.stride * start\n",
      "                ^\n",
      "------------------------------------------------------------\n",
      "\n",
      "pandas/_libs/reduction.pyx:447:16: Assignment to a read-only property\n",
      "\n",
      "Error compiling Cython file:\n",
      "------------------------------------------------------------\n",
      "...\n",
      "        self.buf.shape[0] = length\n",
      "\n",
      "    cdef reset(self):\n",
      "\n",
      "        self.buf.shape[0] = self.orig_len\n",
      "        self.buf.data = self.orig_data\n",
      "                ^\n",
      "------------------------------------------------------------\n",
      "\n",
      "pandas/_libs/reduction.pyx:456:16: Assignment to a read-only property\n",
      "\n",
      "Error compiling Cython file:\n",
      "------------------------------------------------------------\n",
      "...\n",
      "        # move blocks\n",
      "        for i in range(self.nblocks):\n",
      "            arr = self.blocks[i]\n",
      "\n",
      "            # axis=1 is the frame's axis=0\n",
      "            arr.data = self.base_ptrs[i] + arr.strides[1] * start\n",
      "               ^\n",
      "------------------------------------------------------------\n",
      "\n",
      "pandas/_libs/reduction.pyx:573:15: Assignment to a read-only property\n",
      "\n",
      "Error compiling Cython file:\n",
      "------------------------------------------------------------\n",
      "...\n",
      "        # reset blocks\n",
      "        for i in range(self.nblocks):\n",
      "            arr = self.blocks[i]\n",
      "\n",
      "            # axis=1 is the frame's axis=0\n",
      "            arr.data = self.base_ptrs[i]\n",
      "               ^\n",
      "------------------------------------------------------------\n",
      "\n",
      "pandas/_libs/reduction.pyx:592:15: Assignment to a read-only property\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1325, in cythonize_one_helper\n",
      "    return cythonize_one(*m)\n",
      "  File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1301, in cythonize_one\n",
      "    raise CompileError(None, pyx_file)\n",
      "Cython.Compiler.Errors.CompileError: pandas/_libs/reduction.pyx\n",
      "[17/40] Cythonizing pandas/_libs/reshape.pyx\n",
      "[18/40] Cythonizing pandas/_libs/sparse.pyx\n",
      "[19/40] Cythonizing pandas/_libs/testing.pyx\n",
      "[20/40] Cythonizing pandas/_libs/tslib.pyx\n",
      "[21/40] Cythonizing pandas/_libs/tslibs/c_timestamp.pyx\n",
      "[22/40] Cythonizing pandas/_libs/tslibs/ccalendar.pyx\n",
      "[23/40] Cythonizing pandas/_libs/tslibs/conversion.pyx\n",
      "[24/40] Cythonizing pandas/_libs/tslibs/fields.pyx\n",
      "[25/40] Cythonizing pandas/_libs/tslibs/frequencies.pyx\n",
      "warning: pandas/_libs/tslibs/frequencies.pyx:331:36: Unknown type declaration 'bint' in annotation, ignoring\n",
      "warning: pandas/_libs/tslibs/frequencies.pyx:384:38: Unknown type declaration 'bint' in annotation, ignoring\n",
      "[26/40] Cythonizing pandas/_libs/tslibs/nattype.pyx\n",
      "warning: pandas/_libs/tslibs/nattype.pyx:89:0: Global name __nat_unpickle matched from within class scope in contradiction to to Python 'class private name' rules. This may change in a future release.\n",
      "warning: pandas/_libs/tslibs/nattype.pyx:89:0: Global name __nat_unpickle matched from within class scope in contradiction to to Python 'class private name' rules. This may change in a future release.\n",
      "[27/40] Cythonizing pandas/_libs/tslibs/np_datetime.pyx\n",
      "[28/40] Cythonizing pandas/_libs/tslibs/offsets.pyx\n",
      "[29/40] Cythonizing pandas/_libs/tslibs/parsing.pyx\n",
      "warning: pandas/_libs/tslibs/parsing.pyx:742:25: Unknown type declaration 'bint' in annotation, ignoring\n",
      "[30/40] Cythonizing pandas/_libs/tslibs/period.pyx\n",
      "[31/40] Cythonizing pandas/_libs/tslibs/resolution.pyx\n",
      "[32/40] Cythonizing pandas/_libs/tslibs/strptime.pyx\n",
      "[33/40] Cythonizing pandas/_libs/tslibs/timedeltas.pyx\n",
      "\n",
      "Error compiling Cython file:\n",
      "------------------------------------------------------------\n",
      "...\n",
      "        if ((len(number) or len(frac)) and not len(unit)\n",
      "                and current_unit is None):\n",
      "            raise ValueError(\"no units specified\")\n",
      "\n",
      "        if len(frac) > 0 and len(frac) <= 3:\n",
      "            m = 10**(3 -len(frac)) * 1000L * 1000L\n",
      "                                           ^\n",
      "------------------------------------------------------------\n",
      "\n",
      "pandas/_libs/tslibs/timedeltas.pyx:441:43: Cannot assign type 'double' to 'int64_t'\n",
      "\n",
      "Error compiling Cython file:\n",
      "------------------------------------------------------------\n",
      "...\n",
      "            raise ValueError(\"no units specified\")\n",
      "\n",
      "        if len(frac) > 0 and len(frac) <= 3:\n",
      "            m = 10**(3 -len(frac)) * 1000L * 1000L\n",
      "        elif len(frac) > 3 and len(frac) <= 6:\n",
      "            m = 10**(6 -len(frac)) * 1000L\n",
      "                                   ^\n",
      "------------------------------------------------------------\n",
      "\n",
      "pandas/_libs/tslibs/timedeltas.pyx:443:35: Cannot assign type 'double' to 'int64_t'\n",
      "\n",
      "Error compiling Cython file:\n",
      "------------------------------------------------------------\n",
      "...\n",
      "        if len(frac) > 0 and len(frac) <= 3:\n",
      "            m = 10**(3 -len(frac)) * 1000L * 1000L\n",
      "        elif len(frac) > 3 and len(frac) <= 6:\n",
      "            m = 10**(6 -len(frac)) * 1000L\n",
      "        else:\n",
      "            m = 10**(9 -len(frac))\n",
      "                  ^\n",
      "------------------------------------------------------------\n",
      "\n",
      "pandas/_libs/tslibs/timedeltas.pyx:445:18: Cannot assign type 'double' to 'int64_t'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1325, in cythonize_one_helper\n",
      "    return cythonize_one(*m)\n",
      "  File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1301, in cythonize_one\n",
      "    raise CompileError(None, pyx_file)\n",
      "Cython.Compiler.Errors.CompileError: pandas/_libs/tslibs/timedeltas.pyx\n",
      "[34/40] Cythonizing pandas/_libs/tslibs/timestamps.pyx\n",
      "[35/40] Cythonizing pandas/_libs/tslibs/timezones.pyx\n",
      "[36/40] Cythonizing pandas/_libs/tslibs/tzconversion.pyx\n",
      "[37/40] Cythonizing pandas/_libs/window/aggregations.pyx\n",
      "warning: pandas/_libs/window/aggregations.pyx:59:47: Buffer unpacking not optimized away.\n",
      "warning: pandas/_libs/window/aggregations.pyx:59:47: Buffer unpacking not optimized away.\n",
      "warning: pandas/_libs/window/aggregations.pyx:60:47: Buffer unpacking not optimized away.\n",
      "warning: pandas/_libs/window/aggregations.pyx:60:47: Buffer unpacking not optimized away.\n",
      "\n",
      "Error compiling Cython file:\n",
      "------------------------------------------------------------\n",
      "...\n",
      "        buf = <float64_t *>arr.data\n",
      "        bufarr = np.empty(win, dtype=float)\n",
      "        oldbuf = <float64_t *>bufarr.data\n",
      "        for i in range((win - offset), (N - offset)):\n",
      "            buf = buf + 1\n",
      "            bufarr.data = <char *>buf\n",
      "                  ^\n",
      "------------------------------------------------------------\n",
      "\n",
      "pandas/_libs/window/aggregations.pyx:1394:18: Assignment to a read-only property\n",
      "\n",
      "Error compiling Cython file:\n",
      "------------------------------------------------------------\n",
      "...\n",
      "            bufarr.data = <char *>buf\n",
      "            if counts[i] >= minp:\n",
      "                output[i] = func(bufarr, *args, **kwargs)\n",
      "            else:\n",
      "                output[i] = NaN\n",
      "        bufarr.data = <char *>oldbuf\n",
      "              ^\n",
      "------------------------------------------------------------\n",
      "\n",
      "pandas/_libs/window/aggregations.pyx:1399:14: Assignment to a read-only property\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1325, in cythonize_one_helper\n",
      "    return cythonize_one(*m)\n",
      "  File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1301, in cythonize_one\n",
      "    raise CompileError(None, pyx_file)\n",
      "Cython.Compiler.Errors.CompileError: pandas/_libs/window/aggregations.pyx\n",
      "[38/40] Cythonizing pandas/_libs/window/indexers.pyx\n",
      "[39/40] Cythonizing pandas/_libs/writers.pyx\n",
      "[40/40] Cythonizing pandas/io/sas/sas.pyx\n",
      "multiprocessing.pool.RemoteTraceback: \n",
      "\"\"\"\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/lib/python3.7/multiprocessing/pool.py\", line 121, in worker\n",
      "    result = (True, func(*args, **kwds))\n",
      "  File \"/usr/lib/python3.7/multiprocessing/pool.py\", line 44, in mapstar\n",
      "    return list(map(*args))\n",
      "  File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1325, in cythonize_one_helper\n",
      "    return cythonize_one(*m)\n",
      "  File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1301, in cythonize_one\n",
      "    raise CompileError(None, pyx_file)\n",
      "Cython.Compiler.Errors.CompileError: pandas/_libs/reduction.pyx\n",
      "\"\"\"\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"setup.py\", line 757, in <module>\n",
      "    ext_modules=maybe_cythonize(extensions, compiler_directives=directives),\n",
      "  File \"setup.py\", line 542, in maybe_cythonize\n",
      "    return cythonize(extensions, *args, **kwargs)\n",
      "  File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1125, in cythonize\n",
      "    result.get(99999)  # seconds\n",
      "  File \"/usr/lib/python3.7/multiprocessing/pool.py\", line 657, in get\n",
      "    raise self._value\n",
      "Cython.Compiler.Errors.CompileError: pandas/_libs/reduction.pyx\n",
      "Obtaining file:///home/dan/notebooks/pandas-dev/pandas-dev\n",
      "  Preparing metadata (setup.py) ... \u001b[?25lerror\n",
      "  \u001b[1;31merror\u001b[0m: \u001b[1msubprocess-exited-with-error\u001b[0m\n",
      "  \n",
      "  \u001b[31m×\u001b[0m \u001b[32mpython setup.py egg_info\u001b[0m did not run successfully.\n",
      "  \u001b[31m│\u001b[0m exit code: \u001b[1;36m1\u001b[0m\n",
      "  \u001b[31m╰─>\u001b[0m \u001b[31m[156 lines of output]\u001b[0m\n",
      "  \u001b[31m   \u001b[0m Compiling pandas/_libs/reduction.pyx because it changed.\n",
      "  \u001b[31m   \u001b[0m Compiling pandas/_libs/tslibs/timedeltas.pyx because it changed.\n",
      "  \u001b[31m   \u001b[0m Compiling pandas/_libs/window/aggregations.pyx because it changed.\n",
      "  \u001b[31m   \u001b[0m [1/3] Cythonizing pandas/_libs/reduction.pyx\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m Error compiling Cython file:\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m ...\n",
      "  \u001b[31m   \u001b[0m             object cached_typ = None\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m         arr = self.arr\n",
      "  \u001b[31m   \u001b[0m         chunk = self.dummy\n",
      "  \u001b[31m   \u001b[0m         dummy_buf = chunk.data\n",
      "  \u001b[31m   \u001b[0m         chunk.data = arr.data\n",
      "  \u001b[31m   \u001b[0m              ^\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m pandas/_libs/reduction.pyx:101:13: Assignment to a read-only property\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m Error compiling Cython file:\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m ...\n",
      "  \u001b[31m   \u001b[0m                     # On the first pass, we check the output shape to see\n",
      "  \u001b[31m   \u001b[0m                     #  if this looks like a reduction.\n",
      "  \u001b[31m   \u001b[0m                     _check_result_array(res, len(self.dummy))\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m                 PyArray_SETITEM(result, PyArray_ITER_DATA(it), res)\n",
      "  \u001b[31m   \u001b[0m                 chunk.data = chunk.data + self.increment\n",
      "  \u001b[31m   \u001b[0m                      ^\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m pandas/_libs/reduction.pyx:140:21: Assignment to a read-only property\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m Error compiling Cython file:\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m ...\n",
      "  \u001b[31m   \u001b[0m                 PyArray_SETITEM(result, PyArray_ITER_DATA(it), res)\n",
      "  \u001b[31m   \u001b[0m                 chunk.data = chunk.data + self.increment\n",
      "  \u001b[31m   \u001b[0m                 PyArray_ITER_NEXT(it)\n",
      "  \u001b[31m   \u001b[0m         finally:\n",
      "  \u001b[31m   \u001b[0m             # so we don't free the wrong memory\n",
      "  \u001b[31m   \u001b[0m             chunk.data = dummy_buf\n",
      "  \u001b[31m   \u001b[0m                  ^\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m pandas/_libs/reduction.pyx:144:17: Assignment to a read-only property\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m Error compiling Cython file:\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m ...\n",
      "  \u001b[31m   \u001b[0m                 PyArray_SETITEM(result, PyArray_ITER_DATA(it), res)\n",
      "  \u001b[31m   \u001b[0m                 chunk.data = chunk.data + self.increment\n",
      "  \u001b[31m   \u001b[0m                 PyArray_ITER_NEXT(it)\n",
      "  \u001b[31m   \u001b[0m         finally:\n",
      "  \u001b[31m   \u001b[0m             # so we don't free the wrong memory\n",
      "  \u001b[31m   \u001b[0m             chunk.data = dummy_buf\n",
      "  \u001b[31m   \u001b[0m                  ^\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m pandas/_libs/reduction.pyx:144:17: Assignment to a read-only property\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m Error compiling Cython file:\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m ...\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m         self.orig_data = self.buf.data\n",
      "  \u001b[31m   \u001b[0m         self.orig_len = self.buf.shape[0]\n",
      "  \u001b[31m   \u001b[0m         self.orig_stride = self.buf.strides[0]\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m         self.buf.data = self.values.data\n",
      "  \u001b[31m   \u001b[0m                 ^\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m pandas/_libs/reduction.pyx:437:16: Assignment to a read-only property\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m Error compiling Cython file:\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m ...\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m         self.buf.data = self.values.data\n",
      "  \u001b[31m   \u001b[0m         self.buf.strides[0] = self.stride\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m     cdef advance(self, Py_ssize_t k):\n",
      "  \u001b[31m   \u001b[0m         self.buf.data = <char*>self.buf.data + self.stride * k\n",
      "  \u001b[31m   \u001b[0m                 ^\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m pandas/_libs/reduction.pyx:441:16: Assignment to a read-only property\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m Error compiling Cython file:\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m ...\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m     cdef move(self, int start, int end):\n",
      "  \u001b[31m   \u001b[0m         \"\"\"\n",
      "  \u001b[31m   \u001b[0m         For slicing\n",
      "  \u001b[31m   \u001b[0m         \"\"\"\n",
      "  \u001b[31m   \u001b[0m         self.buf.data = self.values.data + self.stride * start\n",
      "  \u001b[31m   \u001b[0m                 ^\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m pandas/_libs/reduction.pyx:447:16: Assignment to a read-only property\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m Error compiling Cython file:\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m ...\n",
      "  \u001b[31m   \u001b[0m         self.buf.shape[0] = length\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m     cdef reset(self):\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m         self.buf.shape[0] = self.orig_len\n",
      "  \u001b[31m   \u001b[0m         self.buf.data = self.orig_data\n",
      "  \u001b[31m   \u001b[0m                 ^\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m pandas/_libs/reduction.pyx:456:16: Assignment to a read-only property\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m Error compiling Cython file:\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m ...\n",
      "  \u001b[31m   \u001b[0m         # move blocks\n",
      "  \u001b[31m   \u001b[0m         for i in range(self.nblocks):\n",
      "  \u001b[31m   \u001b[0m             arr = self.blocks[i]\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m             # axis=1 is the frame's axis=0\n",
      "  \u001b[31m   \u001b[0m             arr.data = self.base_ptrs[i] + arr.strides[1] * start\n",
      "  \u001b[31m   \u001b[0m                ^\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m pandas/_libs/reduction.pyx:573:15: Assignment to a read-only property\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m Error compiling Cython file:\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m ...\n",
      "  \u001b[31m   \u001b[0m         # reset blocks\n",
      "  \u001b[31m   \u001b[0m         for i in range(self.nblocks):\n",
      "  \u001b[31m   \u001b[0m             arr = self.blocks[i]\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m             # axis=1 is the frame's axis=0\n",
      "  \u001b[31m   \u001b[0m             arr.data = self.base_ptrs[i]\n",
      "  \u001b[31m   \u001b[0m                ^\n",
      "  \u001b[31m   \u001b[0m ------------------------------------------------------------\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m pandas/_libs/reduction.pyx:592:15: Assignment to a read-only property\n",
      "  \u001b[31m   \u001b[0m Traceback (most recent call last):\n",
      "  \u001b[31m   \u001b[0m   File \"<string>\", line 36, in <module>\n",
      "  \u001b[31m   \u001b[0m   File \"<pip-setuptools-caller>\", line 34, in <module>\n",
      "  \u001b[31m   \u001b[0m   File \"/home/dan/notebooks/pandas-dev/pandas-dev/setup.py\", line 757, in <module>\n",
      "  \u001b[31m   \u001b[0m     ext_modules=maybe_cythonize(extensions, compiler_directives=directives),\n",
      "  \u001b[31m   \u001b[0m   File \"/home/dan/notebooks/pandas-dev/pandas-dev/setup.py\", line 542, in maybe_cythonize\n",
      "  \u001b[31m   \u001b[0m     return cythonize(extensions, *args, **kwargs)\n",
      "  \u001b[31m   \u001b[0m   File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1134, in cythonize\n",
      "  \u001b[31m   \u001b[0m     cythonize_one(*args)\n",
      "  \u001b[31m   \u001b[0m   File \"/home/dan/venv/pandas-bisect/lib/python3.7/site-packages/Cython/Build/Dependencies.py\", line 1301, in cythonize_one\n",
      "  \u001b[31m   \u001b[0m     raise CompileError(None, pyx_file)\n",
      "  \u001b[31m   \u001b[0m Cython.Compiler.Errors.CompileError: pandas/_libs/reduction.pyx\n",
      "  \u001b[31m   \u001b[0m \u001b[31m[end of output]\u001b[0m\n",
      "  \n",
      "  \u001b[1;35mnote\u001b[0m: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "\u001b[1;31merror\u001b[0m: \u001b[1mmetadata-generation-failed\u001b[0m\n",
      "\n",
      "\u001b[31m×\u001b[0m Encountered error while generating package metadata.\n",
      "\u001b[31m╰─>\u001b[0m See above for output.\n",
      "\n",
      "\u001b[1;35mnote\u001b[0m: This is an issue with the package mentioned above, not pip.\n",
      "\u001b[1;36mhint\u001b[0m: See above for details.\n",
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.0.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m23.2.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[?25hTraceback (most recent call last):\r\n",
      "  File \"/home/dan/notebooks/pandas-dev/pandas-dev/pandas/__init__.py\", line 32, in <module>\r\n",
      "    from pandas._libs import hashtable as _hashtable, lib as _lib, tslib as _tslib\r\n",
      "  File \"/home/dan/notebooks/pandas-dev/pandas-dev/pandas/_libs/__init__.py\", line 3, in <module>\r\n",
      "    from .tslibs import (\r\n",
      "  File \"/home/dan/notebooks/pandas-dev/pandas-dev/pandas/_libs/tslibs/__init__.py\", line 3, in <module>\r\n",
      "    from .conversion import localize_pydatetime, normalize_date\r\n",
      "ModuleNotFoundError: No module named 'pandas._libs.tslibs.conversion'\r\n",
      "\r\n",
      "During handling of the above exception, another exception occurred:\r\n",
      "\r\n",
      "Traceback (most recent call last):\r\n",
      "  File \"t.py\", line 2, in <module>\r\n",
      "    import pandas as pd\r\n",
      "  File \"/home/dan/notebooks/pandas-dev/pandas-dev/pandas/__init__.py\", line 37, in <module>\r\n",
      "    f\"C extension: {module} not built. If you want to import \"\r\n",
      "ImportError: C extension: No module named 'pandas._libs.tslibs.conversion' not built. If you want to import pandas from the source directory, you may need to run 'python setup.py build_ext --inplace --force' to build the C extensions first.\r\n"
     ]
    }
   ],
   "source": [
    "!cd pandas-dev && bash bisect.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0cc1d2ee",
   "metadata": {
    "papermill": {
     "duration": 1.346594,
     "end_time": "2023-05-04T10:12:07.458780",
     "exception": false,
     "start_time": "2023-05-04T10:12:06.112186",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bisecting: a merge base must be tested\n",
      "[d3f08566a80239a18a813ebda9a2ebb0368b1dc5] RLS: 1.0.0rc0\n"
     ]
    }
   ],
   "source": [
    "!cd pandas-dev && bash setup_bisect.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "b4570fdb",
   "metadata": {
    "papermill": {
     "duration": 3494.564339,
     "end_time": "2023-05-04T11:10:22.040086",
     "exception": false,
     "start_time": "2023-05-04T10:12:07.475747",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "output = !cd pandas-dev && git bisect run bash bisect.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2b815d71",
   "metadata": {
    "papermill": {
     "duration": 0.02719,
     "end_time": "2023-05-04T11:10:22.085396",
     "exception": false,
     "start_time": "2023-05-04T11:10:22.058206",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "for line in output:\n",
    "    if 'is the first bad commit' in line:\n",
    "        print(line)\n",
    "        commit = line.split()[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "6fbb033b",
   "metadata": {
    "papermill": {
     "duration": 0.677948,
     "end_time": "2023-05-04T11:10:22.779865",
     "exception": false,
     "start_time": "2023-05-04T11:10:22.101917",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'commit' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_77267/2561882650.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msystem\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf'cd pandas-dev && git log --format=%B -n 1 {commit}'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'commit' is not defined"
     ]
    }
   ],
   "source": [
    "os.system(f'cd pandas-dev && git log --format=%B -n 1 {commit}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0da04363",
   "metadata": {
    "papermill": {
     "duration": 0.027804,
     "end_time": "2023-05-04T11:10:22.826930",
     "exception": false,
     "start_time": "2023-05-04T11:10:22.799126",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "with open('output.txt', 'w') as fd:\n",
    "    fd.write('\\n'.join(output))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00a1067d-621c-44ec-b0ad-1e9c86a82b3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f62902d4-2bf7-4b95-82d7-455332c0febe",
   "metadata": {},
   "outputs": [],
   "source": [
    "GOOD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc0af7a2-a460-4e18-83b7-bbe9dc191aec",
   "metadata": {},
   "outputs": [],
   "source": [
    "BAD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82648a9e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pandas-bisect",
   "language": "python",
   "name": "pandas-bisect"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.17"
  },
  "papermill": {
   "default_parameters": {},
   "duration": 5110.735437,
   "end_time": "2023-05-04T11:10:23.808440",
   "environment_variables": {},
   "exception": null,
   "input_path": "__notebook__.ipynb",
   "output_path": "__notebook__.ipynb",
   "parameters": {},
   "start_time": "2023-05-04T09:45:13.073003",
   "version": "2.4.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
